### PR TITLE
feat(semver): add "semver" module

### PR DESCRIPTION
**Pull request by justjavac**
link #2427

### CODE REVIEW

File: semver/mod.ts [hunk 1]
```diff
@@ -0,0 +1,1806 @@
+// Copyright Isaac Z. Schlueter and Contributors. All rights reserved. ISC license.
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
```
Comment (bartlomieju) [inline semver/mod.ts:4]: Could you add a JSDoc with `@module` with similar description as in README?

**Review by bartlomieju — COMMENTED** [review]
Looks great!

**Review by bartlomieju — APPROVED** [review]
LGTM, @kt3k does this look good to you?

**Review by kt3k — COMMENTED** [review]
Left one comment for now.

@bartlomieju I'll do more detailed review tomorrow

**Comment by justjavac** [comment]
```
deno upgrade --canary --version=76107649804e674268becd693b7b2a954eecb3da
// test ✅
deno upgrade --canary --version=0c87dd1e9898d7ac93e274d3611ee491a107d47a
// test ❎
```

caused by denoland/deno#15118

**Comment by kitsonk** [comment]
ping @dsherret our first issue with using swc for emit...

**Comment by dsherret** [comment]
I will cry myself to sleep tonight.

**Comment by dsherret** [comment]
Ok, it wasn't as bad as I thought. I opened https://github.com/denoland/deno_std/pull/2439

**Review by kt3k — COMMENTED** [review]
@justjavac I have a few questions for some APIs:

- Do we need to support `loose` option? Is it useful for anything? The loose mode seems to increase the complexity of implementation very much. How about dropping the support of this mode?
- I'm also skeptical about the necessity of `clean` and `coerce`. I don't think the behaviors are defined well and I don't find the case where those are useful (Why do you need to handle invalid semver as some valid semver?). How about dropping these APIs?

**Comment by justjavac** [comment]
@kt3k I looked up the commit history and issues of 'node_semver', the `loose` option seems to be used to deal with non-standard npm version numbers before 2013, such as `1.2.3b1`. Now it's 2020, we don't need to include this non-standard semver format in Deno.

I will remove `loose` parameters and `clean` and `coerce` functions.

**Comment by justjavac** [comment]
Ready for review

**Review by kt3k — COMMENTED** [review]
@justjavac Looks much nicer now. Thanks for updating!

I still observe a few variables with `LOOSE` name such as `NUMERICIDENTIFIERLOOSE`, `MAINVERSIONLOOSE`, etc. Are those still relevant?

**Comment by justjavac** [comment]
It is used for range parse, I have deleted the relevant code

**Review by kt3k — APPROVED** [review]
@justjavac Thanks for updating again!

LGTM

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by GerbenRampaart** [comment]
Hurray for this ! semver is a great fit for std 👍 👍

**Pull request merged**